### ISSUE TITLE
 With the same learning rate, the training process of python cannot be reproduced

### ISSUE THREAD

**Issue by toolgood** [dotnet/TorchSharp#952]
With the same learning rate, the training process of python cannot be reproduced.

python code url: https://github.com/thuml/Autoformer
args:  --is_training 1   --root_path ./dataset/ETT-small/   --data_path ETTh1.csv   --model_id ETTh1_96_24   --model Autoformer   --data ETTh1   --features M   --seq_len 96   --label_len 48   --pred_len 24   --e_layers 2   --d_layers 1   --factor 3   --enc_in 7   --dec_in 7   --c_out 7   --des 'Exp'   --itr 1

csharp code url:  https://github.com/toolgood/OtherFiles/tree/master/ToolGood.SoarSky.StockFormer

python output 
Epoch: 1, Steps: 266 | Train Loss: 0.3453052 Vali Loss: 0.5853946 Test Loss: 0.3715304
Epoch: 2, Steps: 266 | Train Loss: 0.2879474 Vali Loss: 0.5856608 Test Loss: 0.3932027

csharp output 
Epoch: 1, Steps: 267 | Train Loss: 0.036781374 Vali Loss: 0.49154732 Test Loss: 0.9598659
Epoch: 2, Steps: 267 | Train Loss: 0.006978075 Vali Loss: 0.5534928 Test Loss: 1.0090411

The training loss of csharp is smaller, but their learning rate value is the same. The Optimizer is Adam. What's wrong?

The complete output is as follows 

python output 
```
train 8521
val 2857
test 2857
        iters: 100, epoch: 1 | loss: 0.2538728
        speed: 0.2749s/iter; left time: 704.0139s
        iters: 200, epoch: 1 | loss: 0.3728751
        speed: 0.0398s/iter; left time: 97.9756s
Epoch: 1 cost time: 35.338929414749146
Epoch: 1, Steps: 266 | Train Loss: 0.3453052 Vali Loss: 0.5853946 Test Loss: 0.3715304
Validation loss decreased (inf --> 0.585395).  Saving model ...
Updating learning rate to 0.0001
        iters: 100, epoch: 2 | loss: 0.2990033
        speed: 0.7406s/iter; left time: 1699.5676s
        iters: 200, epoch: 2 | loss: 0.2835134
        speed: 0.0382s/iter; left time: 83.7662s
Epoch: 2 cost time: 30.412309646606445
Epoch: 2, Steps: 266 | Train Loss: 0.2879474 Vali Loss: 0.5856608 Test Loss: 0.3932027
EarlyStopping counter: 1 out of 3
Updating learning rate to 5e-05
        iters: 100, epoch: 3 | loss: 0.2481640
        speed: 0.7045s/iter; left time: 1429.4565s
        iters: 200, epoch: 3 | loss: 0.2773028
        speed: 0.0400s/iter; left time: 77.1732s
Epoch: 3 cost time: 30.772044897079468
Epoch: 3, Steps: 266 | Train Loss: 0.2533133 Vali Loss: 0.6161127 Test Loss: 0.4347219
EarlyStopping counter: 2 out of 3
Updating learning rate to 2.5e-05
        iters: 100, epoch: 4 | loss: 0.2599970
        speed: 0.7069s/iter; left time: 1246.2746s
        iters: 200, epoch: 4 | loss: 0.2399460
        speed: 0.0390s/iter; left time: 64.8404s
Epoch: 4 cost time: 30.62981414794922
Epoch: 4, Steps: 266 | Train Loss: 0.2327324 Vali Loss: 0.6407704 Test Loss: 0.4355299
EarlyStopping counter: 3 out of 3
Early stopping
```

csharp output
```
train 8521
val 2857
test 2857
        iters: 100, epoch: 1 | loss: 0.022509418
        speed: 0.06384516800000001s/iter; left time: 164.14592692800002s
        iters: 200, epoch: 1 | loss: 0.016443554
        speed: 0.044794280000000006s/iter; left time: 110.68666588000002s
Epoch: 1 cost time: 00:00:13.1986158
Epoch: 1, Steps: 267 | Train Loss: 0.036781374 Vali Loss: 0.49154732 Test Loss: 0.9598659
Validation loss decreased (∞ --> 0.4915473163127899).  Saving model ...
Updating learning rate to 0.0001
        iters: 100, epoch: 2 | loss: 0.0071967947
        speed: 0.11196523900000001s/iter; left time: 257.967910656s
        iters: 200, epoch: 2 | loss: 0.0043902737
        speed: 0.045005988999999996s/iter; left time: 99.193199756s
Epoch: 2 cost time: 00:00:11.9302480
Epoch: 2, Steps: 267 | Train Loss: 0.006978075 Vali Loss: 0.5534928 Test Loss: 1.0090411
EarlyStopping counter: 1 out of 3
Updating learning rate to 5E-05
        iters: 100, epoch: 3 | loss: 0.0030213161
        speed: 0.112114343s/iter; left time: 228.37691669100002s
        iters: 200, epoch: 3 | loss: 0.0026060923
        speed: 0.044858572s/iter; left time: 86.891053964s
Epoch: 3 cost time: 00:00:11.9373947
Epoch: 3, Steps: 267 | Train Loss: 0.0029374957 Vali Loss: 0.50207216 Test Loss: 1.0560036
EarlyStopping counter: 2 out of 3
Updating learning rate to 2.5E-05
        iters: 100, epoch: 4 | loss: 0.0021781202
        speed: 0.112081013s/iter; left time: 198.38339301s
        iters: 200, epoch: 4 | loss: 0.0019224939
        speed: 0.045758267000000005s/iter; left time: 76.41630589s
Epoch: 4 cost time: 00:00:12.0881568
Epoch: 4, Steps: 267 | Train Loss: 0.0020420542 Vali Loss: 0.48704407 Test Loss: 1.0718473
Validation loss decreased (0.4915473163127899 --> 0.4870440661907196).  Saving model ...
Updating learning rate to 1.25E-05
        iters: 100, epoch: 5 | loss: 0.0017565627
        speed: 0.112698757s/iter; left time: 169.38623177099998s
        iters: 200, epoch: 5 | loss: 0.0016746287
        speed: 0.045273742s/iter; left time: 63.519060026s
Epoch: 5 cost time: 00:00:11.9872905
Epoch: 5, Steps: 267 | Train Loss: 0.001695772 Vali Loss: 0.48082343 Test Loss: 1.0862207
Validation loss decreased (0.4870440661907196 --> 0.48082342743873596).  Saving model ...
Updating learning rate to 6.25E-06
        iters: 100, epoch: 6 | loss: 0.0016332312
        speed: 0.113563005s/iter; left time: 140.36387417999998s
        iters: 200, epoch: 6 | loss: 0.0015062408
        speed: 0.045213997s/iter; left time: 51.363100592s
Epoch: 6 cost time: 00:00:12.0541145
Epoch: 6, Steps: 267 | Train Loss: 0.001533716 Vali Loss: 0.4813921 Test Loss: 1.0915599
EarlyStopping counter: 1 out of 3
Updating learning rate to 3.125E-06
        iters: 100, epoch: 7 | loss: 0.0015009291
        speed: 0.113334175s/iter; left time: 109.820815575s
        iters: 200, epoch: 7 | loss: 0.0014873337
        speed: 0.045214798s/iter; left time: 39.291659462s
Epoch: 7 cost time: 00:00:12.0364793
Epoch: 7, Steps: 267 | Train Loss: 0.001453856 Vali Loss: 0.48283124 Test Loss: 1.0993879
EarlyStopping counter: 2 out of 3
Updating learning rate to 1.5625E-06
        iters: 100, epoch: 8 | loss: 0.0015107003
        speed: 0.113228462s/iter; left time: 79.486380324s
        iters: 200, epoch: 8 | loss: 0.0013759754
        speed: 0.045557742s/iter; left time: 27.425760684s
Epoch: 8 cost time: 00:00:12.0286687
Epoch: 8, Steps: 267 | Train Loss: 0.001411993 Vali Loss: 0.48338422 Test Loss: 1.1016366
EarlyStopping counter: 3 out of 3
Early stopping
```

**Comment by toolgood** [comment]
I check 'Dataset_ ETT_ Hour 'class, through comparing the output file with the python output file, it is found that the data is the same. 
csharp out file: Datasets\ETT-small\ETTh1.txt  
python out file: Datasets\ETT-small\ETTh1-python.txt

**Comment by toolgood** [comment]
Adam's construction method is different .
python : betas=(0.9, 0.999)
torchsharp:  double beta1 = 0.9, double beta2 = 0.99
``` python
   def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                 weight_decay=0, amsgrad=False, *, foreach: Optional[bool] = None,
                 maximize: bool = False, capturable: bool = False,
                 differentiable: bool = False, fused: bool = False)
```

``` cs
      public static Adam Adam(IEnumerable<Parameter> parameters, double lr = 0.001, double beta1 = 0.9, double beta2 = 0.99, double eps = 1E-08, double weight_decay = 0.0, bool amsgrad = false, bool maximize = false)

```

**Comment by toolgood** [comment]
When exporting the Conv1d class, python only exports weight, while torchsharp exports weight and bias .

![图片](https://user-images.githubusercontent.com/17310407/224989497-bcbb2a6d-bfc2-4818-b25e-8596cad5aef4.png)

**Comment by toolgood** [comment]
I found that in python language, 'fix seed' and 'shuffle_flag'is set to true, the training loss is unchanged.
In csharp language, it will change.

 The following is the loss value of python's first epoch 
```
[0.6678487062454224, 0.4524787366390228, 0.5046882033348083, 0.5744467973709106, 0.5756551027297974, 0.6845090389251709, 0.3313789665699005, 0.5979511737823486, 0.4439546465873718, 0.4915148913860321, 0.4912724792957306, 0.44404876232147217, 0.5941401124000549, 0.46690425276756287, 0.6090814471244812, 1.0068126916885376, 0.4156982898712158, 0.3781180679798126, 0.5550784468650818, 0.6552767157554626, 0.6352030634880066, 1.4427599906921387, 1.2081501483917236, 0.7207002639770508, 0.40599343180656433, 0.5429306030273438, 0.4320511817932129, 0.5221219062805176, 0.37252292037010193, 0.35188019275665283, 0.41297832131385803, 0.38414788246154785, 0.3689608871936798, 0.27129900455474854, 0.6686843633651733, 0.8658408522605896, 0.5391094088554382, 0.2920566201210022, 0.2807655334472656, 0.6502887010574341, 0.7660796046257019, 0.4651077091693878, 0.4188723862171173, 0.5760363936424255, 0.4098186492919922, 0.294434517621994, 0.24440552294254303, 0.4073072373867035, 0.26075613498687744, 0.2590142786502838, 0.3000604510307312, 0.2929057776927948, 0.41771200299263, 0.35765668749809265, 0.6133162379264832, 0.29294776916503906, 0.3449173867702484, 0.3616315424442291, 0.27434712648391724, 0.2138499766588211, 0.2706383466720581, 0.20618785917758942, 0.24065431952476501, 0.32681170105934143, 0.5601138472557068, 0.2977100610733032, 0.38593775033950806, 0.4157158136367798, 0.34279894828796387, 0.3540883958339691, 0.31798046827316284, 0.2707972228527069, 0.2266538292169571, 0.23011519014835358, 0.22477608919143677, 0.2738594710826874, 0.22672395408153534, 0.24265922605991364, 0.7945233583450317, 0.8990463018417358, 0.4113006591796875, 0.3257884979248047, 0.33384233713150024, 1.099845051765442, 1.1892921924591064, 0.8250901699066162, 0.964194655418396, 0.8799672722816467, 0.42809590697288513, 0.3502717912197113, 0.4817444980144501, 0.24818652868270874, 0.21830463409423828, 0.2263106405735016, 0.3170298635959625, 0.3483511805534363, 0.36470934748649597, 0.3588840365409851, 0.27879244089126587, 0.5831445455551147, 0.9666406512260437, 0.4920082986354828, 0.44189319014549255, 0.19891993701457977, 0.23103778064250946, 0.37664029002189636, 0.2170313447713852, 0.1968984156847, 0.23692196607589722, 0.3488667607307434, 0.3220440745353699, 0.2655176818370819, 0.29385069012641907, 0.26674124598503113, 1.2043795585632324, 2.541353464126587, 1.0078980922698975, 1.185896635055542, 0.794497013092041, 0.6316409707069397, 0.5721228122711182, 0.27433687448501587, 0.3170728385448456, 0.2864919602870941, 0.25786152482032776, 0.2596592903137207, 0.28108665347099304, 0.2575080990791321, 0.3020671308040619, 0.28703898191452026, 0.21795153617858887, 0.22402580082416534, 0.19323404133319855, 0.2757047414779663, 0.35923904180526733, 0.35019615292549133, 0.19880254566669464, 0.25909021496772766, 0.2422090321779251, 0.24169354140758514, 0.31721511483192444, 0.18567219376564026, 0.11441811919212341, 0.367560476064682, 0.2335393875837326, 0.1895754188299179, 0.1796547770500183, 0.2080245018005371, 0.2084677666425705, 0.44700905680656433, 0.3657303750514984, 0.3623068630695343, 0.4344238042831421, 0.23246756196022034, 0.2517438232898712, 0.2519020736217499, 0.33537757396698, 0.47108742594718933, 0.34600046277046204, 0.19487445056438446, 0.16627345979213715, 0.2592513859272003, 0.5416219234466553, 0.2651780843734741, 0.46371760964393616, 0.6027224659919739, 0.5106933116912842, 0.35341835021972656, 0.3839220404624939, 0.6221559047698975, 0.8751319050788879, 0.40322768688201904, 0.5015226006507874, 0.5089287757873535, 0.4524216949939728, 0.9568167328834534, 0.5924668908119202, 0.5467700958251953, 0.5227953195571899, 0.4647204279899597, 0.5175597071647644, 0.43267586827278137, 0.4887745678424835, 0.3528544306755066, 0.2382981926202774, 0.4222632050514221, 0.38494428992271423, 0.62606281042099, 0.4882313013076782, 0.39118853211402893, 0.26585957407951355, 0.190407395362854, 0.29204776883125305, 0.16710373759269714, 0.33027201890945435, 0.2849031984806061, 0.30837059020996094, 0.8162997961044312, 1.1564242839813232, 0.4954749643802643, 0.5177271962165833, 0.7379615902900696, 0.756536602973938, 0.9259446859359741, 0.4499587118625641, 0.4713045060634613, 0.5301445722579956, 0.27989569306373596, 0.3875049352645874, 0.3457621932029724, 0.5626305937767029, 0.7372573018074036, 0.7294213771820068, 0.3920036256313324, 0.4245353639125824, 0.29296284914016724, 0.353023499250412, 0.4516860544681549, 0.270611047744751, 0.3331722319126129, 0.6188443899154663, 0.5211204886436462, 0.47513893246650696, 0.7025971412658691, 0.5502841472625732, 0.648076057434082, 0.3281711935997009, 0.46724212169647217, 0.5355796813964844, 0.2711975872516632, 0.4123609662055969, 0.614719569683075, 0.4130084812641144, 0.7102216482162476, 0.3811601996421814, 0.437820702791214, 0.33510643243789673, 0.22169823944568634, 0.3828783929347992, 0.27428463101387024, 0.24278783798217773, 0.6926048994064331, 0.3296838402748108, 0.23981553316116333, 0.22699052095413208, 0.25159773230552673, 0.3094874322414398, 0.54619300365448, 0.24476361274719238, 0.35738781094551086, 0.3818354904651642, 0.414455384016037, 0.5417948961257935, 0.8593132495880127, 0.41625136137008667, 0.2718959152698517, 0.20129379630088806, 0.19147495925426483, 0.3219831585884094, 0.5117527842521667, 0.43769073486328125, 0.2086063176393509, 0.2227887064218521, 0.21837611496448517, 0.28554409742355347, 0.2423190474510193]
```

**Comment by NiklasGustafsson** [comment]
> Adam's construction method is different .
> python : betas=(0.9, 0.999)
> torchsharp: double beta1 = 0.9, double beta2 = 0.99

I'll make sure to fix that.  In the meantime, just pass in 0.999 explicitly for beta2.

**Comment by NiklasGustafsson** [comment]
The beta2 issue was addressed in PR #957

**Comment by toolgood** [comment]
> > When exporting the Conv1d class, python only exports weight, while torchsharp exports weight and bias .
> 
> The default for `bias` when constructing a Conv1d is `true` in both PyTorch and TorchSharp.


I cannot use 'importsd.py' and 'exportsd.py' for model sharing.
It is more impossible to use the same initial model for the next test.

**Comment by ChengYen-Tang** [comment]
I think to verify this problem, we must first simplify the problem
Perhaps we can use the classical data set to observe
like MNIST

**Comment by toolgood** [comment]
Alas, there are still too many problems with TorchSharp. It is too tiring to test the code of TorchSharp transferred from python. It is better to use python to save your mind. Try TorchSharp again in a few years.

**Comment by ChengYen-Tang** [comment]
> Alas, there are still too many problems with TorchSharp. It is too tiring to test the code of TorchSharp transferred from python. It is better to use python to save your mind. Try TorchSharp again in a few years.

其實不管是 torchsharp 還是 pytorch 他們都是在C++核心引擎上面封裝一個介面，老實說我認為你這個問題是其他原因導致的

**Comment by GeorgeS2019** [comment]
@toolgood 
不要放弃，重新考虑你所采取的假设。

**Comment by GeorgeS2019** [comment]
@NiklasGustafsson 

I think @toolgood just needs a short break before he will be back with more feedback. I hope :-)

==> Do you have any idea how to address, Over expectations?  Is there a better way to manage and communicate expectations at this stage of TorchSharp's progress?
